### PR TITLE
Allow SkyCoords to be converted to Tables

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

Currently it is possible to add `SkyCoord` objects to a `Table` as a column, and it is (often) possible to initialize a `SkyCoord` from a `Table` using `SkyCoord.guess_from_table()`. This pull request would allow a `SkyCoord` object to be converted to a `Table`. The coordinates (together with differentials and obstimes, if present) of the `SkyCoord` would become the `Table` columns, and and any coordinate frame attributes would be recorded in the `Table` metadata.

### CODE REVIEW

**Comment by mhvk** [comment]
I can see the case based on symmetry, but I think if we do it, it makes more sense to me to have a method on `SkyCoord` rather than let `Table(sc)` behave differently from `Table([sc])`. For a `SkyCoord` method, one could imagine also having the option to pass in an existing table, in which case the relevant columns will be added to it.

@taldcroft, what do you think?

**Comment by eerovaher** [comment]
The main reason I didn't implement a `to_table()` function in `SkyCoord` was because I wasn't sure whether such a function should by default return a `Table` or a `QTable`, whereas right now the type of the output is always obvious. That being said I do like the idea of having a function that adds the `SkyCoord` attributes to an existing table, and in that case the type of the input table would determine the type of the output.

**Comment by mhvk** [comment]
I think `QTable` would be most logical given the units involved, but if one has the option to add to the table, then a user can always pass in `sc.to_table(table=Table())`.

**Comment by taldcroft** [comment]
Thanks @eerovaher.  I agree with @mhvk that this method should definitely be on `SkyCoord`. I also think that the output should unambiguously be `QTable`. The columns can have meaningful units, and we are discouraging using `Table` in those cases. Users can always convert to a table, e.g. `Table(sc.to_table())`. (I'd prefer that syntax to what @mhvk said so we don't need a new parameter).

One thing is that this current PR is doing two things at once, and they really should be separated:
1. Updating the `SkyCoordInfo._represent_as_dict` method to handle differentials.
1. Adding a way to use this for creating a table.

Doing (1) is a good idea (and maybe fixes an issue?), but this needs some care and separate unit testing. While it's true that with the current arg defaults it probably has no impact, it would be better to fix serialization to handle differentials in all cases (e.g. writing to ECSV).

After (1) then doing (2) will be straightforward as a new method on `SkyCoord`. I have some questions about the current implementation. One quick thing is about namespacing, which is to ask what if the user wants to add two SkyCoord to an existing table. Should the SkyCoord attributes stored in meta be encapsulated in a dict to avoid potential conflicts?

I also think that you want to use more of the serialization machinery so that vector attributes get put into the table. E.g. if `obstime` is actually an array corresponding to each element of the `SkyCoord` then `obstime` should probably be a table column, not a meta attribute. 

Most of this work is done with `table.serialize.represent_mixins_as_columns`.

**Comment by mhvk** [comment]
:+1: to fixing the long-standing #10226! And probably best to split this in two steps.

A propos including differentials, see https://github.com/astropy/astropy/issues/10226#issue-608694571 for what I tried at the time. We've been trying to make the representation change a bit more reliable, so the problems I mention there may well have gone away.

**Comment by eerovaher** [comment]
The pull request #11750 should fix #10226, and once that gets merged in I can force-push the `SkyCoord.to_table()` method here.

**Comment by eerovaher** [comment]
Here are the commits that allow a `SkyCoord` to be converted to a `QTable`, or to be merged with a `Table`.

I did not make use of the `table.serialize.represent_mixins_as_columns()` function for three reasons. First, that function would have serialized the `SkyCoord` recursively, which is not desirable if the `SkyCoord` contains e.g. a `Time` object as `obstime`. The second reason is that it would have renamed the columns and metadata in the table, e.g. `sc.ra` would have become `'sc_ra'` in the table instead of the more appropriate `'ra'`. The third reason is that it would have cluttered the table metadata with information about the serialization process.

The function `table.operations._merge_table_meta()` does come in handy when a `SkyCoord` is merged with a `Table`. That function offers three options for dealing with metadata conflicts, so it made sense to allow the same options for dealing with column name conflicts too, even if I had previously thought otherwise.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [inline astropy/coordinates/sky_coordinate.py:1022]:
```suggestion
        with it. In the case of a metadata or column name conflict, data from the
        |SkyCoord| are preferred.
```

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [inline astropy/coordinates/sky_coordinate.py:1028]: https://numpydoc.readthedocs.io/en/latest/format.html
```suggestion
        column_conflicts : str {'warn','error','silent'}
```

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [inline astropy/coordinates/sky_coordinate.py:1031]:
```suggestion
        metadata_conflicts : str {'warn','error','silent'}
```

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [inline astropy/coordinates/sky_coordinate.py:1037]: For the return type it is preferred to be explicit
```suggestion
        `~astropy.table.QTable`
```

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [inline astropy/coordinates/sky_coordinate.py:1074]: It would be nice to have more comments in this block of code.
What is each if-block doing?

File: docs/coordinates/skycoord.rst
Comment (nstarman) [inline docs/coordinates/skycoord.rst:1202]: Perhaps this new section should adjoin / be merged with the ``SkyCoord.guess_from_table`` docs.

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [inline astropy/coordinates/tests/test_sky_coord.py:676]: There should also be round-trip tests that ``SkyCoord.guess_from_table(sc.to_table())`` works as expected.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [inline astropy/coordinates/sky_coordinate.py:1092]: Is this done manually in the table code? E.g. the metadata merge handler uses a registry of functions and abstracts a bit over these details. If column_conflicts in the ``astropy.table`` code is changed, that won't automatically be supported here.

**Review by nstarman — COMMENTED** [review]
LGTM. I like that SkyCoord will have both `to` and `from` table methods. I'm not overly familiar with the `astropy.table` code, so I'm sure the relevant maintainers will have a more discerning eye.

File: astropy/coordinates/sky_coordinate.py
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1022]: I'll wait a couple of days before pushing the changes in case there will be more feedback to incorporate.

File: astropy/coordinates/sky_coordinate.py
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1028]: I believe I copied this line from here: https://github.com/astropy/astropy/blob/01ed9fe9eddd04aa97ccf475e5735231c8a88575/astropy/coordinates/sky_coordinate.py#L122
but I can't be entirely sure because there are also quite a few instances of this elsewhere. I can ensure the new docstring conforms to NumPy standards, and a separate pull request should address any examples in the code already present.

File: astropy/coordinates/sky_coordinate.py
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1037]: If that's the convention then it's best to adhere to it.

File: astropy/coordinates/sky_coordinate.py
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1074]: I would say that at least for the blocks between lines 1083 and 1091 the error messages already provide adequate explanations.

File: docs/coordinates/skycoord.rst
Comment (eerovaher) [reply docs/coordinates/skycoord.rst:1202]: That would be the best place, but `SkyCoord.guess_from_table()` does not seem to have any documentation outside its docstring and a couple of change log entries.

Comment (nstarman) [reply docs/coordinates/skycoord.rst:1202]: Good notice. That's certainly not great. Do you want to raise an issue to add documentation on ``guess_from_table``? It would go HERE. In that case, I suggest changing the RST link 

```suggestion
.. _skycoord-to_from-table:
```
Or something similar. Or hold off until ``guess_from_table`` is documented.

File: astropy/coordinates/tests/test_sky_coord.py
Comment (eerovaher) [reply astropy/coordinates/tests/test_sky_coord.py:676]: Round-tripping would certainly be nice, but I'm not sure there should be any expectations. I have the impression that `guess_from_table()` is named the way it is convey that it might not always work.

Comment (nstarman) [reply astropy/coordinates/tests/test_sky_coord.py:676]: For tables generated "in the wild" it is certainly the case that ``SkyCoord.guess_from_table`` might not work, but for (new, not updated) Tables generated by ``SkyCoord.to_table`` I think the expectation is that it MUST work. 
The ``SkyCoord.to_table`` has a very clear prescription for making and naming columns. If round-tripping doesn't work, then  ``SkyCoord.guess_from_table`` needs to be fixed. Implementing such improvements are out of scope of this PR, but I think a roundtrip test is not.

File: astropy/coordinates/sky_coordinate.py
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1092]: The table operations solve column name conflicts by renaming the conflicting columns, and the user can provide a pattern for the renamed columns (see here: https://docs.astropy.org/en/stable/table/operations.html#column-renaming). This function replaces the old column from the table with a new one from the `SkyCoord` instead. This means that `SkyCoord.to_table()` could be easily used for something like updating coordinates in a table after epoch propagation, which is certainly useful. Replacing the very simple code here with something that tries to make use of the machinery in table operations doesn't seem like a good idea.

File: astropy/coordinates/tests/test_sky_coord.py
Comment (eerovaher) [reply astropy/coordinates/tests/test_sky_coord.py:676]: It seems to me that round-tripping can be expected to always work only if `guess_from_table()` can always recognize if its input is a converted `SkyCoord` or not, and that seems quite complicated. It is also possible to be less ambitious and only require round-tripping to work for some subset of `SkyCoord` objects, but specifying such a subset would be outside the scope of this pull request.

File: docs/coordinates/skycoord.rst
Comment (eerovaher) [reply docs/coordinates/skycoord.rst:1202]: If that documentation hasn't been written so far then maybe we shouldn't expect it to be written anytime soon.

Comment (nstarman) [reply docs/coordinates/skycoord.rst:1202]: Well, there wasn't a strong need because it's documented in the docstring. But now that there's a fleshed out SkyCoord <-> Table, and a section on SkyCoord -> Table. I can raise the issue, if you don't want to. No worries.
I did some work on ``guess_from_table`` a couple months back, so I'm happy to assist and/or contribute to writing the section.

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [reply astropy/coordinates/tests/test_sky_coord.py:676]: It's certainly true that some of the features of ``to_table`` are not currently covered by ``guess_from_table`` — eg  obstime. And achieving parity is out of scope here.
If @mhvk / @adrn / @taldcroft agree that no test of ``SkyCoord.guess_from_table(sc.to_table()) == sc`` should be in this PR than an [issue](https://github.com/astropy/astropy/issues/new?assignees=&labels=&template=feature_request.md) should be made to *start* these tests. Regardless, I do think a [new issue](https://github.com/astropy/astropy/issues/new?assignees=&labels=&template=feature_request.md) is merited to highlight the myriad ways in which ``SkyCoord.guess_from_table`` needs to be improved so that ``SkyCoord.guess_from_table(sc.to_table()) == sc`` works. I can make this issue, or you can. It's your PR and good work here, so the option's yours.

Comment (eerovaher) [reply astropy/coordinates/tests/test_sky_coord.py:676]: I think I overestimated how complicated making round-tripping work would be. If a table has been produced by `SkyCoord.to_table()` then its metadata will contain `frame` and `representation_type`. It is possible to look up what the parameters of a `SkyCoord` with this `frame` and `representation_type` combination are called and looking up which of those are actually present in the table is very simple.  Any possible coordinate frame parameters can be listed with `frame_transform_graph.frame_attributes.keys()`, so any metadata keys not listed there are not relevant for reconstructing the coordinate frame. No guesswork required. So I agree that it is reasonable to expect that _any_ `SkyCoord` could be round-tripped and I no longer think that implementing it would be overly complicated.

Having established that it should be possible to round-trip any `SkyCoord` it becomes apparent that testing such round-tripping would be analogous to testing round-tripping a `SkyCoord` as a mixin column in `Table` serialization (because that is supposed to work for any `SkyCoord` too). So the already existing tests just need to be adapted for the round-tripping here and there shouldn't be any need to write tests from scratch. But I still believe adapting the aforementioned tests would naturally be a part of the PR that would update `guess_from_table()` and not this PR.

File: docs/coordinates/skycoord.rst
Comment (eerovaher) [reply docs/coordinates/skycoord.rst:1202]: I you are willing to write that documentation then it does make sense to rename the RST link.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1092]: All this is great. My original concern is simply that the functionality of column conflicts is mirrored, but not any code itself. Improvements/extensions elsewhere would need to be manually duplicated here. Now this isn't a huge deal on a per-case basis — the per-case marginal cost is relatively low. I'm not an `astropy.table` maintainer and I'm not sure if there even is existing machinery, but my inclination would be to, like metadata in `astropy.utils.metadata`, have a system to handle column conflicts. One code used in multiple places. If this machinery doesn't exist, my question is moot (but might spawn a new issue as that machinery would be very out of scope).

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [reply astropy/coordinates/tests/test_sky_coord.py:676]: > I think I overestimated how complicated making round-tripping work would be.

The best type of estimation. 👍 

> No guesswork required. So I agree that it is reasonable to expect that any SkyCoord could be round-tripped...

Great. A good example of [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle).

> testing such round-tripping would be analogous to testing round-tripping a SkyCoord as a mixin column in Table serialization (because that is supposed to work for any SkyCoord too). So the already existing tests just need to be adapted for the round-tripping here and there shouldn't be any need to write tests from scratch.

Even better. 

> But I still believe adapting the aforementioned tests would naturally be a part of the PR that would update ``guess_from_table()`` and not this PR.

The above shows you've thought deeply on how this can be implemented and if there's a pre-v5.0 timeline for ensuring round-tripping, then I have no objections. My initial hesitation / insistence was predicated on the uncertainty in the amount of effort required to get this all to work. Making a test here would highlight the myriad ways in which ``guess_from_table()`` doesn't yet support round-tripping. But if this is already quantified then I have no objections.

File: astropy/coordinates/sky_coordinate.py
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1092]: The way column name conflicts are handled here is completely different from how the table operations do it. The table joining functions rename the conflicting columns so that the resulting table contains both. The user can specify a pattern for this renaming, but there is no equivalent of the `column_conflicts` argument at all.

The way column name conflicts are handled here has much more in common with how metadata conflicts are handled, but trying to make use of that machinery for columns sounds quite convoluted.

If the user wishes to make use of the machinery in the table joining functions then it is possible to first convert the `SkyCoord` to a table and then combine the tables using the existing functions.

Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1092]: > there is no equivalent of the column_conflicts argument at all.

Ok. Then this code is AOK. when perusing the code I saw the similarity and wondered.

*Resolved*

**Comment by eerovaher** [comment]
I made two very small changes and now both the code and documentation should be good enough to be merged in. However the commit history has become rather messy. Should I squash?

**Comment by mhvk** [comment]
Sorry to come back to this after a *long* time, but I wonder whether there aren't too many features built in here. I really like the creation of the table base on the SkyCoord, and how simple that ends up being re-using `info`. But I am much less sure whether it should include also working with an existing table, especially since it ends up just creating a new table and then merging it with the existing one. I feel it would make more sense to just point people to [`table.hstack`](https://docs.astropy.org/en/stable/table/operations.html#id6).

Ping also @taldcroft.

**Comment by eerovaher** [comment]
There are two differences between this and `table.hstack()` The first is that `hstack()` allows the input tables to have different numbers of rows. The second is that in case of a column name conflict `ħstack()` renames columns to ensure that all columns are present in the output, whereas this here replaces any conflicting columns in a table with those from the `SkyCoord`. As I mentioned in one of the comments this makes it very convenient to perform epoch propagation of coordinates listed in a table. I found that to be simple and useful enough for me to justify implementing it, but it could be that my use case is too specific and is not worth supporting in the core `astropy` package, which is an opinion I am willing to accept if more people express it.

**Comment by taldcroft** [comment]
@mhvk @eerovaher - I thought for a bit about the question of allowing a `table` argument. I see both points, but at the end it seems like allowing the `table` argument (and friends) is a good thing. At least for me it seems like a putting the coord into an existing table would be similarly common to making a standalone table. It's true that one can document how to do the merge, except the case of overwriting existing columns which is basically not possible with `hstack`.  (As an aside, using `join_type='exact'` will check the lengths).

I'm all about limiting complexity, but in this case the complexity lives on its own in an isolated function that isn't really used by the rest of the core. This is not likely to cause a maintenance burden down the line.

And though it is not the most compelling reason, @eerovaher and @nstarman have already done all the work to put this in...

**Comment by mhvk** [comment]
OK, clearly this is showing I should have read the discussion better! I guess this is more like a `Table.update()` (which perhaps should exist, but once it does, this could always be simplified, so no problem there). Anyway, happy to go along with the majority view here.

**Comment by eerovaher** [comment]
Now that you mention `Table.update()` by name I actually think that it would make more sense to implement that and to limit `SkyCoord.to_table()` so that it only converts the `SkyCoord`. The user could then decide if they wish to merge the resulting table using `update()` or `hstack()` or something else.

Do you agree that it would be a good idea to open a separate pull request for `Table.update()`?

**Comment by taldcroft** [comment]
What's the idea for `Table.update()`?

**Comment by eerovaher** [comment]
The `Table.update()` I have in mind would be analogous to Python's `dict.update()`, i.e. `t1.update(t2)` would add columns from `t2` to `t1` or replace columns from `t1` in case of matching column names, but it would of course also merge metadata. So in terms of what has been implemented here the current `sc.to_table(t)` would be the same as `t.update(sc.to_table())`.

**Comment by taldcroft** [comment]
@eerovaher - interesting! 

I wouldn't jump to making a new Table method lightly but if @mhvk agrees then maybe this is a reasonable idea. It's illuminating to look at the pandas DataFrame `update()` method, which is a bit more complicated than the dict analog. This may be more than we need, but worth thinking about.

An alternative would be to add an `overwrite` kwarg to `hstack`. Your example would look like:
```
t = table.hstack([t, sc.to_table()], overwrite=True)
```
This would stack tables from the left and overwrite any existing ones. The upside is you don't need to invent a new method. The downside is that this makes a new table, though in some cases that might be a feature.

**Comment by eerovaher** [comment]
There is now a separate pull request in progress to implement `Table.update()`, so there isn't much reason to include that functionality here. I will force-push a version of `SkyCoord.to_table()` that only performs the conversion.

**Comment by eerovaher** [comment]
9318f427c41198489299d5831d29ba4835b664f7 passed all the tests except the one that checks for the change log entry, but that was added by 5dd8c058adceadd59fb9238fb5628dcab4c8cd8f. This pull request should be ready for merging.

**Comment by nstarman** [comment]
@mhvk. Not to throw a wrench in the works, but this PR is what got me thinking about ``to/from_format()`` in #11559, and I'm wondering if that might also be the right approach here. I can easily imagine the following being useful:

- ``to/from_mapping``
- ``to/from_table``
- ``to/from_array``  (when structured quantities happen)
- ``to/from_nddata`` ?
- ``to/from_string`` (this would nicely simplify the SkyCoord constructor)

To avoid name pollution, Cosmology will be implementing ``obj = cosmo_instance.to_format('[format]')``, ``cosmo_instance = Cosmology.from_format(obj)``. An identical pattern would work here ``sc.to_format('[format]')``, ``SkyCoord.from_format(obj)``. Not much code would need to change in this PR... if that's the way to go.

**Comment by eerovaher** [comment]
Requesting review from @mhvk and @taldcroft.

File: docs/table/construct_table.rst
Comment (taldcroft) [inline docs/table/construct_table.rst:18]: The docs here refer to data types that can be passed to the `Table` constructor as the first argument. So you should revert this change here since you cannot do `Table(SkyCoord(...))`.

File: docs/table/index.rst
Comment (taldcroft) [inline docs/table/index.rst:77]:
```suggestion
adding columns or rows incrementally, even converting from a
:class:`pandas.DataFrame` or a |SkyCoord|.
```

**Review by taldcroft — COMMENTED** [review]
From the perspective of this proposed API, the code looks fine to me with just the minor comments on documentation.

I'm neutral or slightly 👎 on going to `to_format` here. It just seems like more complexity than is justified by the use cases. But I won't argue too strongly apart from wanting to just get a new (nice but somewhat minor) feature merged and move on.

File: docs/table/construct_table.rst
Comment (eerovaher) [reply docs/table/construct_table.rst:18]: There are sections about converting a `SkyCoord` and a `pandas.DataFrame` in the text below, so they should also be mentioned here, but it would indeed be better to clearly distinguish between what can be directly passed to the `Table` constructor and what can be converted with separate functions.

Comment (taldcroft) [reply docs/table/construct_table.rst:18]: I see the first sentence of the `Examples` section is redundant from above anyway. Let's just make this bit simpler with:
```
Examples
=======

The examples below demonstrate a variety of ways to create a |Table| or |QTable|,
either by using the table class constructor or by converting other objects to a table.
```

Comment (eerovaher) [reply docs/table/construct_table.rst:18]: The preceding section already says that examples will follow, and whether creating a table involves the constructor explicitly or is done through a separate function is obvious from the relevant example. So it seems to me this new proposed sentence is not really needed either and the examples can simply start with Setup.

**Comment by eerovaher** [comment]
@taldcroft, are you happy with the documentation changes?

**Review by taldcroft — APPROVED** [review]
@eerovaher - This looks good to me now. Sorry for the long delay in review... We probably need a coordinates maintainer to approve since this is mostly in that package. @mhvk?

**Review by mhvk — APPROVED** [review]
@eerovaher - yes, this really has turned out very nice, and also from me apologies for letting it drag on so long. Will merge now. Thanks again!

**Comment by astrofrog** [comment]
@eerovaher - could you consider opening a pull request to the v5.0.x branch to add an entry to this file:

https://github.com/astropy/astropy/blob/v5.0.x/docs/whatsnew/5.0.rst

to advertise this new feature?

**Pull request merged**
